e2e/txpool_test.go

```diff
@@ -0,0 +1,125 @@
+package e2e
+
+import (
+	"context"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/0xPolygon/minimal/crypto"
+	"github.com/0xPolygon/minimal/e2e/framework"
+	"github.com/0xPolygon/minimal/helper/tests"
+	txpoolOp "github.com/0xPolygon/minimal/txpool/proto"
+	"github.com/0xPolygon/minimal/types"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTxPool_TransactionCoalescing(t *testing.T) {
+	// Test scenario:
+	// Add tx with nonce 0
+	// -> Check if tx has been parsed
+	// Add tx with nonce 2
+	// -> tx shouldn't be executed, but shelved for later
+	// Add tx with nonce 1
+	// -> check if both tx with nonce 1 and tx with nonce 2 are parsed
+
+	// Predefined values
+	gasPrice := big.NewInt(10000)
+
+	referenceKey, referenceAddr := tests.GenerateKeyAndAddr(t)
+	defaultBalance := framework.EthToWei(10)
+
+	devInterval := 5
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(devInterval)
+		config.Premine(referenceAddr, defaultBalance)
+	})
+	srv := srvs[0]
+	client := srv.JSONRPC()
+
+	// Required default values
+	signer := crypto.NewEIP155Signer(100)
+
+	// TxPool client
+	clt := srv.TxnPoolOperator()
+	toAddress := types.StringToAddress("1")
+	oneEth := framework.EthToWei(1)
+
+	generateTx := func(nonce uint64) *types.Transaction {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    nonce,
+			From:     referenceAddr,
+			To:       &toAddress,
+			GasPrice: gasPrice,
+			Gas:      1000000,
+			Value:    oneEth,
+			V:        1, // it is necessary to encode in rlp
+		}, referenceKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		return signedTx
+	}
+
+	generateReq := func(nonce uint64) *txpoolOp.AddTxnReq {
+		var msg *txpoolOp.AddTxnReq
+		unstakeTxn := generateTx(nonce)
```

Comment:
`unstakeTxn`? 😕

Reply:
Whoops, it was leftover :(

Fixed 👍 

a6edca6